File: FilterMapFold.hs
filter' :: (a -> Bool) -> [a] -> [a]
filter' _ [] = []
filter' p (x:xs)
  | p x       = x : filter' p xs
  | otherwise = filter' p xs

filter'' :: (a -> Bool) -> [a] -> [a]
filter'' p l = [x | x <- l, p x]


map' :: (a -> b) -> [a] -> [b]
map' _ []     = []
map' f (x:xs) = f x : map' f xs

map'' :: (a -> b) -> [a] -> [b]
map'' f l = [f x | x <- l]


foldr' :: (a -> b -> b) -> b -> [a] -> b
foldr' _ v []     = v
foldr' f v (x:xs) = f x $ foldr' f v xs

foldl' :: (a -> b -> a) -> a -> [b] -> a
foldl' _ v []     = v
foldl' f v (x:xs) = foldl' f (f v x) xs

File: filter_map_fold.ml
let rec filter' p = function
  | [] -> []
  | x :: xs when p x -> x :: filter' p xs
  | _ :: xs -> filter' p xs;;

let rec map' f = function
  | [] -> []
  | x :: xs -> f x :: map' f xs;;

let rec foldr' f v = function
  | [] -> v
  | x :: xs -> f x (foldr' f v xs);;

let rec foldl' f v = function
  | [] -> v
  | x :: xs -> foldl' f (f v x) xs;;
